src/rgw/rgw_reshard.cc

```diff
@@ -267,79 +268,183 @@ int RGWBucketReshard::set_resharding_status(rgw::sal::RGWRadosStore* store,
+       " target index shard objects: " << cpp_strerror(ret) << dendl;
     return ret;
   }
 
-  cls_rgw_bucket_instance_entry instance_entry;
-  ret = store->getRados()->bucket_set_reshard(bucket_info, instance_entry);
+  if (ret = fault.check("set_target_layout");
+      ret == 0) { // no fault injected, write the bucket instance metadata
+    ret = store->getRados()->put_bucket_instance_info(bucket_info, false,
+                                                      real_time(), nullptr);
+  }
+
   if (ret < 0) {
-    ldout(store->ctx(), 0) << "RGWReshard::" << __func__ <<
-      " ERROR: error setting bucket resharding flag on bucket index: " <<
-      cpp_strerror(-ret) << dendl;
+    ldout(store->ctx(), 0) << "ERROR: " << __func__ << " failed to write "
+        "target index layout to bucket info: " << cpp_strerror(ret) << dendl;
+    // delete the target shard objects (ignore errors)
+    store->svc()->bi->clean_index(bucket_info, *target);
     return ret;
   }
-
   return 0;
-}
+} // init_target_layout
 
-int RGWBucketReshard::clear_index_shard_reshard_status(rgw::sal::RGWRadosStore* store,
-						       const RGWBucketInfo& bucket_info)
+// delete the bucket index shards associated with the target layout and remove
+// it from the bucket instance metadata
+static int revert_target_layout(rgw::sal::RGWRadosStore* store,
+                                RGWBucketInfo& bucket_info,
+                                const ReshardFaultInjector& fault)
 {
-  uint32_t num_shards = bucket_info.layout.current_index.layout.normal.num_shards;
+  auto& layout = bucket_info.layout;
+  auto prev = layout; // make a copy for cleanup
 
-  if (num_shards < std::numeric_limits<uint32_t>::max()) {
-    int ret = set_resharding_status(store, bucket_info,
-				    cls_rgw_reshard_status::NOT_RESHARDING);
-    if (ret < 0) {
-      ldout(store->ctx(), 0) << "RGWBucketReshard::" << __func__ <<
-	" ERROR: error clearing reshard status from index shard " <<
-	cpp_strerror(-ret) << dendl;
-      return ret;
-    }
+  // remove target index shard objects
+  int ret = store->svc()->bi->clean_index(bucket_info, *layout.target_index);
+  if (ret < 0) {
+    ldout(store->ctx(), 1) << "WARNING: " << __func__ << " failed to remove "
+        "target index with: " << cpp_strerror(ret) << dendl;
+    ret = 0; // non-fatal error
   }
 
+  // clear target_index and resharding state
+  layout.target_index = std::nullopt;
+  layout.resharding = rgw::BucketReshardState::None;
+
+  if (ret = fault.check("revert_target_layout");
+      ret == 0) { // no fault injected, revert the bucket instance metadata
+    ret = store->getRados()->put_bucket_instance_info(bucket_info, false, real_time(), nullptr);
+  }
+
+  if (ret < 0) {
+    ldout(store->ctx(), 0) << "ERROR: " << __func__ << " failed to clear "
+        "target index layout in bucket info: " << cpp_strerror(ret) << dendl;
+
+    bucket_info.layout = std::move(prev); // restore in-memory layout
```

Comment:
after restoring the in memory layout, I think we have to commit it.

Reply:
this block is handling the error from `put_bucket_instance_info()` above, where we failed to commit the updated `bucket_info` to rados. so it's just reverting to what we had before